File: src/main/scala/com/vc/hard/L411.scala
package com.vc.hard

object L411 {
  def minAbbreviation(target: String, dictionary: Array[String]): String = {
    import scala.collection.mutable

    //Tries structure to store all the elements from dictionary
    case class Node(child: mutable.HashMap[Char, Node], var isWord: Boolean)
    val root = Node(new mutable.HashMap[Char, Node](), false)
    def add(root: Node, word: String): Unit = {
      var current = root
      word.foreach(ch => {
        if(!current.child.contains(ch)) {
          current.child.put(ch, new Node(new mutable.HashMap[Char, Node](), false))
        }
        current = current.child(ch)
      })
      current.isWord = true
    }
    def search(word: String, current: Node, num: Int): Boolean = {
      if(num > 0) {
        //Keep on skipping characters until num > 0
        current.child.foreach(x => {
          if(search(word, x._2, num - 1)) return true
        })
        return false
      }
      else {
        if(word.length == 0) return current.isWord
        var idx = 0
        var numVar = num
        while(idx < word.length && word(idx) >= '0' && word(idx) <= '9') {
          numVar = numVar * 10 + word(idx) - '0'
          idx += 1
        }
        if(idx != 0) {
          //Word starts with number
          return search(word.substring(idx), current, numVar)
        }
        else {
          //Word start with characters
          val ch = word(0)
          if(current.child.contains(ch))
            return search(word.substring(1), current.child(ch), numVar)
          else
            return false
        }
      }
    }


    //Get abbreviation of given length for target word
    val sb = new mutable.StringBuilder()
    val list = new mutable.ListBuffer[String]()
    def getAbbr(currentIndex: Int, requiredLength: Int): Unit = {
      if(currentIndex >= target.length) return
      val prevNum = sb.nonEmpty && sb(sb.length - 1) >= '0' && sb(sb.length - 1) <= '9'

      if(requiredLength == 1) {
        if(currentIndex == target.length - 1) {
          sb.append(target(currentIndex))
          list += sb.toString
        }
        else if(!prevNum) {
          val remainingLengthFromTarget = target.length - currentIndex
          sb.append(remainingLengthFromTarget)
          list += sb.toString
        }
      }
      else {
        val endIndex = sb.length

        //Append letter
        sb.append(target(currentIndex))
        getAbbr(currentIndex + 1, requiredLength - 1)
        sb.delete(endIndex, sb.length)

        //Append number
        if(!prevNum) {
          (currentIndex + 1 until target.length).foreach(i => {
            sb.append(i - currentIndex)
            getAbbr(i, requiredLength - 1)
            sb.delete(endIndex, sb.length)
          })
        }
      }
    }

    //Add all elements from dict into tries for searching
    dictionary.foreach(word => {
      add(root, word)
    })

    //Start with index 1 & go until target.length
    (1 until target.length).foreach(i => {
      list.clear
      sb.clear
      getAbbr(0, i)
      println(list.mkString(", "))
      list.foreach(word => {
        if(!search(word, root, 0)) return word
      })
    })
    target
  }
}

File: src/main/java/com/vc/hard/L411.java
package com.vc.hard;

import java.util.*;

class L411 {
    class Node {
        HashMap<Character, Node> map;
        boolean isWord;

        Node() {
            map = new HashMap<>();
        }
    }

    class Trie {
        Node root = new Node();
        private void addWord(String word) {
            Node current = root;
            for(int i = 0; i < word.length(); i++) {
                char ch = word.charAt(i);
                if(!current.map.containsKey(ch)) current.map.put(ch, new Node());
                current = current.map.get(ch);
            }
            current.isWord = true;
        }

        private boolean search(String abbr, Node current, int index, int num) {
            //System.out.println("abbr: " +abbr + " index: "+index+" num: "+num);
            if(num > 0) {
                for(Map.Entry<Character, Node> entry: current.map.entrySet()) {
                    if(search(abbr, entry.getValue(), index, num - 1)) return true;
                }
                return false;
            }
            else if(index == abbr.length()) return current.isWord;
            else {
                int oldIndex = index;
                while(index < abbr.length() && abbr.charAt(index) >= '0' && abbr.charAt(index) <= '9') {
                    num = num * 10 + abbr.charAt(index) - '0';
                    index++;
                }
                if(index != oldIndex) {
                    return search(abbr, current, index, num);
                }
                while(index < abbr.length()) {
                    char ch = abbr.charAt(index);
                    if(current.map.containsKey(ch)) return search(abbr, current.map.get(ch), index + 1, num);
                    index++;
                }
                return false;
            }
        }
    }

    public String minAbbreviation(String target, String[] dictionary) {
        Trie trie = new Trie();
        for(String str: dictionary) {
            trie.addWord(str);
        }

        List<String> list = new ArrayList<>();
        for(int i = 1; i <= target.length(); i++) {
            list.clear();
            getAbbr(target, "", 0, i, false, list);
            for(String l: list) if(!trie.search(l, trie.root, 0, 0)) return l;
            //System.out.println("target: "+target+" length: "+i);
            //System.out.println(list);
        }
        return "";
    }

    private void getAbbr(String word,
                         String abbr, int index, int requiredLength, boolean prevNum, List<String> res) {
        // System.out.println("word: "+word+" abbr: "+abbr+" index: "
        //                    +index+" requiredLength: "+requiredLength+" prevNum: "+prevNum);
        if(index >= word.length()) return;
        if(!prevNum) {
            String remaining = (word.length() - index) + "";
            res.add(abbr + remaining);
        }
        if(requiredLength == 1) {
            if(index == word.length() - 1) res.add(abbr + word.charAt(index));
        }
        else {
            getAbbr(word, abbr + word.charAt(index), index + 1, requiredLength - 1, false, res);
            if(!prevNum) {
                for(int i = index + 1; i < word.length(); i++) {
                    getAbbr(word, abbr + (i - index), i, requiredLength - 1, true, res);
                }
            }
        }
    }
}
